### PR TITLE
Reporter:  allow extending samples with extra attributes

### PR DESCRIPTION
**Pull request by athre0z**
This PR does two things:

### Replace `addProfileAttributes` and `attrKeyValue[T]` with `AttrTableManager`

I found the previous design with the generic on `addProfileAttributes` rather inelegant. Firstly, there's no covariance on Go generics, so it would take two `addProfileAttributes` calls and an `append` to add both string and int attributes:

https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/47e8410f9a1516a03c79a1f390c0c0bdd316921b/reporter/otlp_reporter.go#L669-L675

It also required reflection in `addProfileAttributes` for no good reason. IMHO this isn't really a good use-case for generics. In `AttrTableManager` this is replaced with two separate methods that each add just one attribute.

Secondly, a proper type for this results in a much cleaner interface for the second change discussed below.

### Introduce `SampleAttrProducer` interface

This allows agent implementations to add custom sample attributes without maintaining a fork of the reporter implementation. It can gather extra meta-data when the trace is initially added and then emit extra attributes when the sample is sent out.

### CODE REVIEW

File: reporter/attrmgr_test.go
Comment (athre0z) [inline reporter/attrmgr_test.go:22]: Not sure why this was in the struct: it seems that simply making this a local temporary variable in the loop below is easier. I reworked it accordingly, but perhaps I'm missing something on why this was done like this originally.

File: reporter/attrmgr_test.go
Comment (athre0z) [inline reporter/attrmgr_test.go:32]: Same here.

File: reporter/config.go
Comment (florianl) [inline reporter/config.go:66]: Should this be moved to `iface.go`?

File: reporter/attrmgr.go
Comment (florianl) [inline reporter/attrmgr.go:24]: `*profiles.Profile` is internal to the package reporter. How are users of the package supposed to create it?
How should the integration look like, for a different kind of reporter, like for the OTel collector?  see https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/208

From the updated testcase (`TestGetSampleAttributes`), where this API is used, it looks more like this API build a parallel `*profiles.Profile` on top of an existing one.

File: reporter/attrmgr.go
Comment (florianl) [inline reporter/attrmgr.go:37]: This opens the discussion again around sending empty values for attributes that are not required - see https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/233

Comment (rockdaboot) [inline reporter/attrmgr.go:37]: How would skipping empty strings work here (as suggested in #233)? We can not just return a 0 index.

File: reporter/attrmgr.go
Comment (rockdaboot) [inline reporter/attrmgr.go:1]: Add vanity import
```suggestion
package reporter // import "go.opentelemetry.io/ebpf-profiler/reporter"
```

File: reporter/attrmgr.go
Comment (rockdaboot) [inline reporter/attrmgr.go:31]: Use type `attribute.Key` for key, so the string conversions can be removed from the caller side. (In the future we should use semconv keys only.)

File: reporter/attrmgr.go
Comment (rockdaboot) [inline reporter/attrmgr.go:45]: Simplification: You can build the compoundKey inside the function with
```
	compoundKey := fmt.Sprintf("%v_%v", key, value)
```

File: reporter/attrmgr.go
Comment (rockdaboot) [inline reporter/attrmgr.go:24]: We can generalize the AttrTableManager with
```
func NewAttrTableManager(attrTable *[]*common.KeyValue) *AttrTableManager {
```

Comment (athre0z) [reply reporter/attrmgr.go:24]: How is this private? It's a public type from the OTEL proto lib?

> From the updated testcase (TestGetSampleAttributes), where this API is used, it looks more like this API builds a parallel *profiles.Profile on top of an existing one just with attributes but without traces.

No, it's just populating the attribute table portion. I'll apply the change that Tim suggested [here](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/237#discussion_r1839868496) to make this clearer.

File: reporter/attrmgr.go
Comment (athre0z) [reply reporter/attrmgr.go:45]: I originally considered that, but there's a bunch of private protobuf fields in `AnyValue` that will make the key unnecessarily long. It also ends up printing pointers, so it is prone to break if Go were to switch to a moving GC. Example formatting of `AnyValue` object:

```
{{{} [] [] <nil>} 0 [] 0xc00011f610}{{{} [] [] <nil>} 0 [] 0xc00011f630}{{{} [] [] <nil>} 0 [] 0xc00011f660}
```

File: reporter/attrmgr.go
Comment (athre0z) [reply reporter/attrmgr.go:24]: Yeah, probably makes things a bit clearer. I went with the original approach to simplify letting the manager also maintain `attribute_units` if we want to populate that in the future, but I suppose we can just pass a pointer to that as well then.

File: reporter/config.go
Comment (athre0z) [reply reporter/config.go:66]: I considered that originally, but `iface.go` is currently dedicated to the public interface implemented by any reporter. If found `config.go` to be a better fit. I didn't want readers to incorrectly conclude that they have to implement this for a reporter.

File: reporter/attrmgr.go
Comment (athre0z) [reply reporter/attrmgr.go:34]: I was under the impression that the discussion in #233 was leaning towards not doing the filtering in the function/component responsible for managing the attribute table. In the interest of not spreading the discussion over two issues, I'll let this PR wait until #233 is either merged. If we decide on letting the manager fn filter, we can simply rework this to:

```go
AppendStringSttr(attrs []AttrIndex, key, value) []AttrIndex
```

File: reporter/attrmgr.go
Comment (athre0z) [reply reporter/attrmgr.go]: d0a6a5b

File: reporter/attrmgr.go
Comment (athre0z) [inline reporter/attrmgr.go:22]: Alternatively we could not take that any argument, and rename the type to `AttrTableBuilder`. In that case I'd then add a new method `GetAttrTable()` that returns the final built table. That'd change the workflow for the caller to:

- Create empty instance
- Populate it with `AddXXX` calls
- Take the final table with `GetAttrTable()` and manually assign it to the attribute table field

Perhaps more obvious than mutating the slice field in place through a pointer? No strong preference here.

Comment (rockdaboot) [reply reporter/attrmgr.go:22]: I thought about this as well (named it `Get()` to not repeat AttrTable), but it comes with the downside that you need to call this function when done. And this can be forgotten, while the current solution is safer from accidental failure.

My not very strong preference would be to avoid an extra `Get` function call.

File: reporter/attrmgr.go
Comment (rockdaboot) [reply reporter/attrmgr.go:43]: When using value.GetValue() it looks like
```
service.name_&{apmServiceName2}
process.pid_&{6789}
```
Not sure why this should change with a moving GC.

File: reporter/otlp_reporter.go
Comment (rockdaboot) [inline reporter/otlp_reporter.go:608]: nit: Rename `AddStringAttr` to `AddString` to avoid repetition of "Attr". Same for `AddIntAttr`.

File: reporter/attrmgr.go
Comment (athre0z) [reply reporter/attrmgr.go:43]: Ah yeah, with `GetValue` it looks differently, since `AnyValue_StringValue` and `AnyValue_IntValue` don't have the private protobuf meta fields. Looks like they aren't generated for the `oneof` field structs (probably reuse cache in parent).

I can add what you suggest, but honestly I don't think it makes the code any better, since it relies on a lot of implicit assumptions. If we were to add support for `AnyValue_ArrayValue` (that contains a pointer) later, it'd be unsafe again because it will compare pointers instead of contents, with even the same instance not being guaranteed to be considered equal if a moving GC was introduced.

So to save a grand total of 2 lines of code, we'd now have introduced a bunch of possible future bugs.

Comment (rockdaboot) [reply reporter/attrmgr.go:43]: Not sure how likely `AnyValue_ArrayValue` becomes added here, but feel free to stay with the current code. It has the advantage that we the option to optimize building the compound string when Sprintf turns out to be a slow horse (variadic arguments always escape to the heap last time I tested).

**Review by rockdaboot — APPROVED** [review]

File: reporter/config.go
Comment (florianl) [inline reporter/config.go:69]: Can the returned `any` replaced with a (generic) type?
Maybe generic type parameters could help here? My worry is, that `any` might introduce subtle bugs by not using this API correctly.

Comment (athre0z) [reply reporter/config.go:69]: Not elegantly. It'd require us to also propagate the type parameter to `Config` and `OTLPReporter` structs.

File: reporter/otlp_reporter.go
Comment (athre0z) [reply reporter/otlp_reporter.go]: Done, though name additionally changed further to `AppendOptionalString` because I switched to an append-based approach to support optional attributes.

File: reporter/attrmgr.go
Comment (athre0z) [reply reporter/attrmgr.go]: Implemented this now, but with pointer arg (`attrs *[]AttrIndex`) that is mutated in place instead of returning the extended slice.

**Review by rockdaboot — APPROVED** [review]

File: reporter/config.go
Comment (florianl) [reply reporter/config.go:69]: My personal take is, that `any`/`interfacer{}` should be avoided as return values in part of public APIs. As things are moving and this API is not considered stable, I think we can take this as an opportunity and improve in a next step.

**Review by florianl — APPROVED** [review]

File: reporter/config.go
Comment (athre0z) [reply reporter/config.go:69]: I generally agree with avoiding `any`, but in this case I don't think that there's any elegant way to use generics here. There's no covariance for different instances of a generic interface. That is, we can't e.g. take `SampleAttrProducer[any]` as an argument and then fill it with a `SampleAttrProducer[SomethingConcrete]`. So the generic argument will be "infectious" and propagate up. Having potentially multiple generic arguments on both the config and the reporter type seems worse than having a very localized `any`.

**Pull request merged**
